lib/kube/proxy/websocket_client_test.go

```diff
@@ -241,7 +323,111 @@ func (e *wsStreamExecutor) stream(conn *gwebsocket.Conn, options clientremotecom
 	return err
 }
 
-func (e *wsStreamExecutor) connectViaWebsocket() error {
+const (
+	// portforwardDataChan is the websocket channel for transferring data.
+	portforwardDataChan = iota
+	// portforwardErrChan is the websocket channel for transferring errors.
+	portforwardErrChan
+)
+
+// portforward opens a listener at the specified port and waits for new connections.
+// once a connection is received, it waits for the upstream data and writes it
+// into the original connection.
+// Due to portforward websocket limitations, the listener do not accept
+// concurrent connections.
+func (e *wsStreamClient) portforward(remoteConn *gwebsocket.Conn) (err error) {
+	if e.localPort == nil || e.readyChan == nil {
+		return trace.Errorf("cannot use portforward without proper initialization")
+	}
+
+	e.listener, err = net.Listen("tcp", fmt.Sprintf("localhost:%d", *e.localPort))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	close(e.readyChan)
+	for {
+		wg := sync.WaitGroup{}
+		errChan := make(chan error, 3)
+		conn, err := e.listener.Accept()
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				err = nil
+			}
+			return trace.Wrap(err)
+		}
+
+		wg.Add(2)
+
+		go func() {
+			defer wg.Done()
+			p := make([]byte, 1024)
+			for {
+				n, err := conn.Read(p)
+				if err != nil {
+					if errors.Is(err, io.EOF) {
+						err = nil
+					}
+					errChan <- trace.Wrap(err)
+					return
+				}
+				if err := remoteConn.WriteMessage(
+					gwebsocket.BinaryMessage,
+					append([]byte{portforwardDataChan}, p[:n]...),
+				); err != nil {
+					errChan <- trace.Wrap(err)
+					return
+				}
+			}
+		}()
+
+		go func() {
+			defer wg.Done()
+			for {
+				_, buf, err := remoteConn.ReadMessage()
+				// if len(buf)==3, it is the mandatory protocol message. This message
+				// includes the port number from proxy and we can safelly ignore it.
+				// [channel, uint16]
+				if len(buf) > 1 && len(buf) != 3 {
+					// We let the server send the stream number and we choose the desired stream accordingly.
+					// If the stream is nil, we ignore the payload and continue.
+					switch buf[0] {
+					case portforwardDataChan:
+						_, err := conn.Write(buf[1:])
+						if err != nil {
+							errChan <- trace.Wrap(err)
+							return
+						}
+					case portforwardErrChan:
+						err := trace.Errorf(string(buf[1:]))
+						errChan <- trace.Wrap(err)
+						// Once we receive an error from streamErr, we must stop processing.
+						// The server also stops the execution and closes the connection.
+						return
+					}
+				}
+
+				if err != nil {
+					// check the connection was properly closed by server, and if true ignore the error.
+					var websocketErr *gwebsocket.CloseError
+					if errors.As(err, &websocketErr) && websocketErr.Code == gwebsocket.CloseNormalClosure {
+						err = nil
+					}
+					errChan <- trace.Wrap(err)
+					return
+				}
+			}
+		}()
+
+		wg.Wait()
+		close(errChan)
+		err = <-errChan
+		if err != nil {
```

Comment:
This assumes that either the first error is non-nil, or all errors are nil. Is that correct? Could we write down the assumption in a comment?